saltstack/base/salt/userdata-secrets/bin/retrieve-userdata-secrets.sh

```diff
@@ -0,0 +1,68 @@
+#!/bin/bash
+
+set -e +x
+
+# This function implements basically the same exponential backoff strategy as the aws cli in standard retry mode
+retry_until_non_placeholder_value() {
+  local BASE_FACTOR=2
+  local MAX_BACKOFF=20
+  local MAX_ATTEMPT=15
+
+  local total_wait_time=0
+  local timeout=1
+  local attempt=1
+
+  SECRET_STRING="$("$@")"
+  while [[ "$SECRET_STRING" == "PLACEHOLDER" ]]; do
+    if (( attempt > MAX_ATTEMPT )); then
+      echo "Maximum number of retry attempts reached while waiting for userdata secret to update! Total wait time was $total_wait_time seconds. Exiting..."
+      exit 1
+    fi
+
+    echo "Userdata secret still has PLACEHOLDER value. Waiting $timeout seconds until the next attempt to retrieve the secret."
+    sleep $timeout
+
+    total_wait_time=$(( total_wait_time + timeout ))
+    attempt=$(( attempt + 1))
+    timeout=$(( timeout * BASE_FACTOR > MAX_BACKOFF ? MAX_BACKOFF : timeout * BASE_FACTOR ))
+
+    SECRET_STRING="$("$@")"
+  done
+  echo "Userdata secret's value was retrieved after waiting $total_wait_time seconds."
+}
+
+# Intermittent issues like throttling of AWS APIs are handled by the built in retry mechanism of the aws cli.
+# Waiting for CB to update the secret's value to the actual secrets is handled by the function above.
+main() {
+  if [[ "$CLOUD_PLATFORM" == "AWS" ]]; then
+    # See: https://docs.aws.amazon.com/cli/latest/userguide/cli-configure-retries.html
+    # aws cli with standard retry mode uses an exponential backoff with a base factor of 2 and a maximum backoff time of 20 seconds
```

Comment:
Great stuff! Just thinking out loud, the standard retry strategy does not account for access denied errors (presumably 403 or 400, however this is mapped in the CLI response).

This will hopefully not be an issue; the secret resource has initially full access (no resource policy), they the resource policy is set (permitting access to a specific EC2 instance), followed by setting the real secret value. In practice, no access denied shall be encountered when executed from the target EC2 instance, so let us ignore this until if surfaces (if ever). :smile:

Reply:
Should hopefully be fine, since we atomically update the resource policy and the secret value at once.